src/shogun/clustering/KMeansBase.cpp

```diff
@@ -0,0 +1,405 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 1999-2008 Gunnar Raetsch
+ * Written (W) 2007-2009 Soeren Sonnenburg
+ * Copyright (C) 1999-2009 Fraunhofer Institute FIRST and Max-Planck-Society
+ */
+
+#include <shogun/clustering/KMeansBase.h>
+#include <shogun/distance/Distance.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/base/Parallel.h>
+
+#ifdef HAVE_PTHREAD
+#include <pthread.h>
+#endif
+
+#define MUSRECALC
+
+using namespace shogun;
+
+CKMeansBase::CKMeansBase()
+: CDistanceMachine()
+{
+	init();
+}
+
+CKMeansBase::CKMeansBase(int32_t k_, CDistance* d, bool use_kmpp)
+: CDistanceMachine()
+{	
+	init();
+	k=k_;
+	set_distance(d);
+	use_kmeanspp=use_kmpp;
+}
+
+CKMeansBase::CKMeansBase(int32_t k_i, CDistance* d_i, SGMatrix<float64_t> centers_i)
+: CDistanceMachine()
+{
+	init();
+	k = k_i;
+	set_distance(d_i);
+	set_initial_centers(centers_i);
+}
+
+CKMeansBase::~CKMeansBase()
+{
+}	
+
+void CKMeansBase::set_initial_centers(SGMatrix<float64_t> centers)
+{	
+	
+	dimensions = ((CDenseFeatures<float64_t>*) distance->get_lhs())->get_num_features();
+
+	REQUIRE(centers.num_cols == k,
+			"Expected %d initial cluster centers, got %d", k, centers.num_cols);
+
+	REQUIRE(centers.num_rows == dimensions,
+			"Expected %d dimensionional cluster centers, got %d", dimensions, centers.num_rows);
+
+	m_mus_initial = centers;
+}
+
+void CKMeansBase::set_random_centers(SGVector<float64_t> weights_set, SGVector<int32_t> cl_list, int32_t XSize)
+{
+	CDenseFeatures<float64_t>* lhs=
+		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
+
+	for (int32_t i=0; i<XSize; i++)
+	{
+		const int32_t Cl=CMath::random(0, k-1);
+		weights_set[Cl]+=1.0;
+		cl_list[i]=Cl;
+
+		int32_t vlen=0;
+		bool vfree=false;
```

Comment:
could this work with the new SGVector methods? rather than bool do_free and pointers?